runbotics/runbotics-scheduler/src/migrations/1733815803333-OldLogsManagement.ts

```diff
@@ -0,0 +1,149 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class OldLogsManagement1733815803333 implements MigrationInterface {
+    name = 'OldLogsManagement1733815803333';
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`
+            CREATE OR REPLACE PROCEDURE public.remove_old_process_instance(batch_limit integer DEFAULT 1, cutoff_date date DEFAULT (CURRENT_DATE - '6 months'::interval))
+            LANGUAGE plpgsql
+            AS $procedure$
+            DECLARE
+                rows_deleted INT;
+                rows_left INT;
+                half_year_back DATE;
+            BEGIN
+                half_year_back := CURRENT_DATE - INTERVAL '6 months';
+
+                IF cutoff_date > CURRENT_DATE - INTERVAL '1 months' THEN
+                    RAISE EXCEPTION 'cutoff_date must be at least 1 months before the current date: %', cutoff_date;
+                END IF;
+
+                RAISE NOTICE 'Starting deletion of records older than %.', cutoff_date;
+                RAISE NOTICE 'Omitting those processes that have less than 10 runs since % (last 6 months).', half_year_back;
+
+                LOOP
+                    BEGIN
+                    DELETE FROM process_instance
+                    WHERE id IN (
+                        SELECT id
+                        FROM process_instance
+                        WHERE created < cutoff_date
+                            AND process_id IN (
+                                SELECT process_id
+                                FROM process_instance
+                                WHERE created > half_year_back
+                                GROUP BY process_id
+                                HAVING COUNT(id) > 10
+                            )
+                        LIMIT batch_limit
+                    );
```

Comment:
what is the difference between `cutoff_date` and `half_year_back`?

Reply:
`cutoff_date` is given by the user and is `<today_minus_6_months>` by default, but can be set to any date. Logs after that date are supposed to be deleted.
`half_year_back` is always `<today_minus_6_months>`, is used to check how many runs the process does have in the last half year